src/main/java/com/google/sps/servlets/RecipeServlet.java

```diff
@@ -73,121 +81,129 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     String data = request.getReader().lines().collect(Collectors.joining(System.lineSeparator()));
     Recipe newRecipe = gson.fromJson(data, Recipe.class);
-    if (newRecipe.content == null || newRecipe.title == null) {
+    newRecipe.metadata = gson.fromJson(data, RecipeMetadata.class);
+    if (newRecipe.content == null || newRecipe.metadata.title == null) {
       response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
       return;
     }
 
     String token = request.getParameter("token");
-    FirebaseToken decodedToken = Auth.verifyIdToken(token);
-    if (decodedToken == null) {
-      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+    String uid = getUid(token, response);
+    if (uid == null) {
       return;
     }
 
-    DocumentReference recipeRef = DBUtils.recipes().document();
-    newRecipe.id = recipeRef.getId();
-    newRecipe.creatorId = decodedToken.getUid();
-    ApiFuture addRecipeFuture = recipeRef.set(newRecipe);
+    newRecipe.metadata.creatorId = uid;
+    String recipeId = dbInterface.addRecipe(newRecipe.metadata, newRecipe.content);
```

Comment:
Feel free to change this method in `DBInterface` and `FirestoreDB` to simply take a `Recipe` object.

Reply:
Part of me wants to leave that for another PR just because of how big this already is

Reply:
Eh, I'd say it fits here.